src/storage/txn/sched_pool.rs

```diff
@@ -102,94 +183,103 @@ impl SchedPool {
                 } else {
-                    worker_pool.spawn(f)
+                    fail_point!("single_queue_pool_task");
+                    self.vanilla.as_ref().unwrap().spawn(priority_level, f)
                 }
             }
-            SchedPool::Priority {
-                worker_pool,
-                resource_ctl,
-            } => {
-                let fixed_level = match priority {
-                    CommandPri::High => Some(0),
-                    CommandPri::Normal => None,
-                    CommandPri::Low => Some(2),
-                };
-                // TODO: maybe use a better way to generate task_id
-                let task_id = rand::random::<u64>();
-                let mut extras = Extras::new_multilevel(task_id, fixed_level);
-                extras.set_metadata(group_name.as_bytes().to_owned());
-                worker_pool.spawn_with_extras(
-                    ControlledFuture::new(
-                        async move {
-                            f.await;
-                        },
-                        resource_ctl.clone(),
-                        group_name.as_bytes().to_owned(),
-                    ),
-                    extras,
-                )
-            }
         }
     }
 
     pub fn scale_pool_size(&self, pool_size: usize) {
-        match self {
-            SchedPool::Vanilla {
-                high_worker_pool,
-                worker_pool,
-            } => {
-                high_worker_pool.scale_pool_size(std::cmp::max(1, pool_size / 2));
-                worker_pool.scale_pool_size(pool_size);
+        match self.queue_type {
+            QueueType::Vanilla => {
+                let vanilla = self.vanilla.as_ref().unwrap();
+                vanilla.scale_pool_size(pool_size);
+            }
+            QueueType::Priority => {
+                let priority = self.priority.as_ref().unwrap();
+                priority.scale_pool_size(pool_size);
             }
-            SchedPool::Priority { worker_pool, .. } => {
-                worker_pool.scale_pool_size(pool_size);
+            QueueType::Dynamic => {
+                let vanilla = self.vanilla.as_ref().unwrap();
+                let priority = self.priority.as_ref().unwrap();
+                priority.scale_pool_size(pool_size);
+                vanilla.scale_pool_size(pool_size);
             }
         }
     }
 
-    pub fn get_pool_size(&self, priority: CommandPri) -> usize {
-        match self {
-            SchedPool::Vanilla {
-                high_worker_pool,
-                worker_pool,
-            } => {
-                if priority == CommandPri::High {
-                    high_worker_pool.get_pool_size()
+    fn can_use_priority(&self) -> bool {
+        match self.queue_type {
+            QueueType::Vanilla => false,
+            QueueType::Priority => true,
+            QueueType::Dynamic => self.priority.as_ref().unwrap().resource_ctl.is_customized(),
```

Comment:
if that and vanilla pool is empty, can we just drop the vanilla?

Reply:
what do you mean? the vanilla pool should not be empty in the dynamic type.

Reply:
Empty means no tasks are being handled in the pool

Reply:
Do you want to reduce threads in this way, it's in a hot path, so if we need to do it, maybe like before way https://github.com/tikv/tikv/pull/14582/commits/d84fb00c33cf170e8c7671cf2c92a181ac8eec92 ?
But as @glorv said, I think we kept it is ok, this isn't a big problem here.

Reply:
But it is still possible the switch between this two pool according to the user's resource group config. And since this `SchedPool` is read pool, we can't change any field either.